src/java/main/org/apache/zookeeper/server/util/LogChopper.java

```diff
@@ -0,0 +1,152 @@
+        BinaryInputArchive logStream = BinaryInputArchive.getArchive(is);
+        BinaryOutputArchive choppedStream = BinaryOutputArchive.getArchive(os);
+        FileHeader fhdr = new FileHeader();
+        fhdr.deserialize(logStream, "fileheader");
+
+        if (fhdr.getMagic() != FileTxnLog.TXNLOG_MAGIC) {
+            System.err.println("Invalid magic number in txn log file");
+            return false;
+        }
+        System.out.println("ZooKeeper Transactional Log File with dbid "
+                + fhdr.getDbid() + " txnlog format version "
+                + fhdr.getVersion());
+
+        fhdr.serialize(choppedStream, "fileheader");
+        int count = 0;
+        boolean hasZxid = false;
+        long previousZxid = -1;
+        while (true) {
+            long crcValue;
+            byte[] bytes;
+            try {
+                crcValue = logStream.readLong("crcvalue");
+
+                bytes = logStream.readBuffer("txnEntry");
+            } catch (EOFException e) {
+                System.out.println("EOF reached after " + count + " txns.");
+                // returning false because nothing was chopped
+                return false;
+            }
+            if (bytes.length == 0) {
+                // Since we preallocate, we define EOF to be an
+                // empty transaction
+                System.out.println("EOF reached after " + count + " txns.");
+                // returning false because nothing was chopped
+                return false;
+            }
+
+            Checksum crc = new Adler32();
+            crc.update(bytes, 0, bytes.length);
+            if (crcValue != crc.getValue()) {
+                throw new IOException("CRC doesn't match " + crcValue +
+                        " vs " + crc.getValue());
+            }
+            TxnHeader hdr = new TxnHeader();
+            Record txn = SerializeUtils.deserializeTxn(bytes, hdr);
+            if (logStream.readByte("EOR") != 'B') {
+                System.out.println("Last transaction was partial.");
+                throw new EOFException("Last transaction was partial.");
+            }
+
+            long txnZxid = hdr.getZxid();
+            if (txnZxid == zxid) {
+                hasZxid = true;
+            }
+
+            // logging the gap to make the inconsistency investigation easier
+            if (previousZxid != -1 && txnZxid != previousZxid + 1) {
+                long txnEpoch = ZxidUtils.getEpochFromZxid(txnZxid);
+                long txnCounter = ZxidUtils.getCounterFromZxid(txnZxid);
+                long previousEpoch = ZxidUtils.getEpochFromZxid(previousZxid);
+                if (txnEpoch == previousEpoch || txnCounter != 1) {
+                    System.out.println(
+                        String.format("There is gap between %x and %x",
+                                previousZxid, txnZxid));
+                }
+            }
+            previousZxid = txnZxid;
+
+            if (hdr.getZxid() > zxid) {
```

Comment:
`hdr.getZxid()` is previously assigned to `txnZxid`, so maybe use `txnZxid` here instead? 

Also, would it be better to move this check before the gap detection logic to terminate earlier?

Reply:
I'd like to leave this check after the gap detection. This covers the one scenario where there is a gap in the log that the target zxid just happens to fall in. Otherwise the operator will only learn that the target is not in the log, not that/why the log is corrupt.